src/java/org/apache/cassandra/index/sai/disk/v2/sortedbytes/SortedBytesWriter.java

```diff
@@ -0,0 +1,192 @@
+            final int suffixLength = term.length - prefixLength;
+            assert suffixLength > 0; // terms are unique
+
+            bytesOutput.writeByte((byte) (Math.min(prefixLength, 15) | (Math.min(15, suffixLength - 1) << 4)));
+            if (prefixLength >= 15)
+            {
+                bytesOutput.writeVInt(prefixLength - 15);
+            }
+            if (suffixLength >= 16)
+            {
+                bytesOutput.writeVInt(suffixLength - 16);
+            }
+            bytesOutput.writeBytes(term.bytes, term.offset + prefixLength, term.length - prefixLength);
+        }
+        maxLength = Math.max(maxLength, term.length);
+        previous.copyBytes(term);
+
+        pointId++;
+    }
+
+    public static class Meta
+    {
+        public final long trieFP;
+        public final long count;
+        public final int maxTermLength;
+        public final byte[] offsetMetaBytes;
+        public final long offsetBlockCount;
+
+        public Meta(long trieFP, long count, int maxTermLength, byte[] offsetMetaBytes, long offsetBlockCount)
+        {
+            this.trieFP = trieFP;
+            this.count = count;
+            this.maxTermLength = maxTermLength;
+            this.offsetMetaBytes = offsetMetaBytes;
+            this.offsetBlockCount = offsetBlockCount;
+        }
+    }
+
+    public Meta finish() throws IOException
+    {
+        long offsetBlockCount = (pointId + TERMS_DICT_BLOCK_MASK) >>> TERMS_DICT_BLOCK_SHIFT;
+
+        final PackedLongValues offsets = offsetsBuilder.build();
+
+        RAMOutputStream meta = new RAMOutputStream();
+
+        // DirectMonotonicReader is much faster than MonotonicBlockPackedReader
+        DirectMonotonicWriter offsetsWriter = DirectMonotonicWriter.getInstance(meta, offsetsOutput, offsetBlockCount, DIRECT_MONOTONIC_BLOCK_SHIFT);
+        for (int x = 0; x < offsetBlockCount; x++)
+        {
+            offsetsWriter.add(offsets.get(x));
+        }
+        offsetsWriter.finish();
+
+        byte[] metaBytes = new byte[(int)meta.getFilePointer()];
+        meta.writeTo(metaBytes, 0);
+
+        final long trieFP = this.trieWriter.complete();
+        return new Meta(trieFP, pointId, maxLength, metaBytes, offsetBlockCount);
+    }
+
+    public static int gatherBytes(ByteComparable term, BytesRefBuilder builder)
+    {
+        return gatherBytes(term.asComparableBytes(ByteComparable.Version.OSS41), builder);
+    }
+
+    public static int gatherBytes(ByteSource byteSource, BytesRefBuilder builder)
+    {
+        builder.clear();
+        int length = 0;
+        // gather the term bytes from the byteSource
+        while (true)
+        {
+            final int val = byteSource.next();
+            if (val != ByteSource.END_OF_STREAM)
+            {
+                ++length;
+                builder.append((byte) val);
+            }
+            else
+            {
+                break;
+            }
+        }
```

Comment:
Can be simplified to:
```
        int val;
        // gather the term bytes from the byteSource
        while ((val = byteSource.next()) != ByteSource.END_OF_STREAM)
        {
            ++length;
            builder.append((byte) val);
        }
```

Reply:
No thanks, it's from Lucene and it is easier to see that with fewer non performance related changes.